src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,162 @@
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> oldNode = getElementByIndex(index);
+        T oldValue = oldNode.value;
+        oldNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> currentNode = getElementByIndex(index);
+        unlink(currentNode);
+        return currentNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = first;
+        while (currentNode != null) {
+            if (currentNode.value == object
+                    || currentNode.value != null
+                    && currentNode.value.equals(object)) {
+                unlink(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean checkValidIndex(int index) {
+        return index < 0 || index > size;
+    }
+
+    private void addFirst(T value) {
+        if (first == null && last == null) {
+            first = new Node<>(value, null, null);
+            last = first;
+        } else {
+            Node<T> newElement = new Node<>(value, null, first);
+            first.previous = newElement;
+            first = newElement;
+        }
+        size++;
+    }
+
+    private void addLast(T value) {
+        if (first == null && last == null) {
+            first = new Node<>(value, null, null);
+            last = first;
+        } else {
+            Node<T> newElement = new Node<>(value, last, null);
+            last.next = newElement;
+            last = newElement;
+        }
+        size++;
+    }
+
+    private Node<T> getElementByIndex(int index) {
+        if (index == size || checkValidIndex(index)) {
+            throw new IndexOutOfBoundsException("Exception");
+        }
+        Node<T> result;
+        if (index < size / 2) {
+            result = first;
+            for (int i = 0; i < index; i++) {
+                result = result.next;
+            }
+        } else {
+            result = last;
+            for (int i = size - 1; i > index; i--) {
+                result = result.previous;
+            }
+        }
+        return result;
+    }
+
+    private void changeLink(Node<T> node, T value) {
+        Node<T> nextNode = node;
+        Node<T> previousNode = nextNode.previous;
+        Node<T> newNode = new Node<>(value, previousNode, nextNode);
+        nextNode.previous = newNode;
+        previousNode.next = newNode;
+        size++;
+    }
+
+    private void unlink(Node<T> node) {
+        if (size == 1) {
+            first = null;
+            last = null;
+        } else {
+            if (node.equals(first)) {
```

Comment:
or another option to check is it first node: `node.previous == null` (can be left as is)